source/slang/slang-dataflow.h

```diff
@@ -0,0 +1,148 @@
+#pragma once
+
+#include "../core/slang-basic.h"
+
+#include <algorithm>
+#include <limits>
+#include <queue>
+
+namespace Slang
+{
+
+template<typename D>
+struct InOutPair
+{
+    D in;
+    D out;
+};
+
+//
+// Parameters:
+// - context: This object provides the transfer function
+// - ns: The list of nodes for which we should return the eventual output of
+//       their transfer function. The closure of predecessor nodes must be
+//       finite. Should not contain duplicates.
+// - strictlyDependsOn:
+//      A function f(x, y) which returns true if x depends on the output of y
+//      and y does not depend on x (and thus y should be evaluated before x)
+// - maxNumUpdates:
+//      (optional) the upper limit for how many times to evaluate the
+//      transfer function at each node.
+//
+// Returns:
+// - A list of eventual inputs and outputs of the transfer functions of the
+//   nodes specified in `ns`
+//
+template<typename T, typename P>
+List<InOutPair<typename T::Domain>> dataFlow(
+    T& context,
+    List<typename T::Node> ns,
+    P strictlyDependsOn,
+    typename T::Domain initialInput = T::Domain::bottom(),
+    UInt maxNumUpdates = std::numeric_limits<UInt>::max())
+{
+    using Node = typename T::Node;
+    using Domain = typename T::Domain;
+    struct Info
+    {
+        Domain in;
+        Domain out;
+        bool dirty;
+        UInt numUpdates;
+    };
+
+    // If we can't evaluate our transfer function, return the only possible value
+    if(maxNumUpdates == 0)
+        return List<InOutPair<Domain>>::makeRepeated(InOutPair<Domain>{initialInput, Domain::bottom()}, ns.getCount());
+
+    // Our algorithm state, `infos` tracks the input output of each node as
+    // well as if it should be recomputed
+    // TODO: We could prune known-stable nodes from this dictionary as we go...
+    Dictionary<Node, Info> infos;
+    std::priority_queue<Node, std::vector<Node>, P> workQueue(strictlyDependsOn);
```

Comment:
Is there any reason to use priority queue? It seems like a normal work list is sufficient.

Reply:
We've trying to avoid using std containers in this codebase, unless it is strictly necessary.